src/libraries/System.Private.CoreLib/src/System/Delegate.cs

```diff
@@ -57,6 +57,68 @@ public abstract partial class Delegate : ICloneable, ISerializable
 
         public virtual Delegate[] GetInvocationList() => new Delegate[] { this };
 
+        /// <summary>
+        /// Gets a value that indicates whether the <see cref="Delegate"/> has a single invocation target.
+        /// </summary>
+        /// <value>true if the <see cref="Delegate"/> has a single invocation target.</value>
+        public bool HasSingleTarget => Unsafe.As<MulticastDelegate>(this).HasSingleTarget;
+
+        /// <summary>
+        /// Gets an enumerator for the invocation targets of this delegate.
+        /// </summary>
+        /// <remarks>
+        /// This returns a <see cref="InvocationListEnumerator{TDelegate}"/>" /> that follows the IEnumerable pattern and
+        /// thus can be used in a C# 'foreach' statements to retrieve the invocation targets of this delegate without allocations.
+        /// The order of the delegates returned by the enumerator is the same order in which the current delegate invokes the methods that those delegates represent.
+        /// The method returns an empty enumerator for null delegate.
+        /// </remarks>
+        public static System.Delegate.InvocationListEnumerator<TDelegate> EnumerateInvocationList<TDelegate>(TDelegate? d) where TDelegate : System.Delegate
+            => new InvocationListEnumerator<TDelegate>(Unsafe.As<MulticastDelegate>(d));
+
+        /// <summary>
+        /// Provides an enumerator for the invocation list of a delegate.
+        /// </summary>
+        /// <typeparam name="TDelegate">Delegate type being enumerated.</typeparam>
+        public struct InvocationListEnumerator<TDelegate> where TDelegate : System.Delegate
+        {
+            private readonly MulticastDelegate? _delegate;
+            private int _index;
+
+            internal InvocationListEnumerator(MulticastDelegate? d)
+            {
+                _delegate = d;
+                _index = -1;
+            }
+
+            /// <summary>
+            /// Implements the IEnumerator pattern.
+            /// </summary>
+            public TDelegate Current
+            {
+                get => Unsafe.As<TDelegate>(_delegate?.TryGetAt(_index));
+            }
+
+            /// <summary>
+            /// Implements the IEnumerator pattern.
+            /// </summary>
+            public bool MoveNext()
+            {
+                int index = _index + 1;
+                if (_delegate?.TryGetAt(index) == null)
```

Comment:
I assume it's cheaper to have MoveNext do the TryGetAt and then have Current do the TryGetAt again than to have an extra _current TDelegate field that MoveNext sets and Current returns?

Reply:
Good point. It is cheaper for empty enumerator; about the same for 1 target; and more expensive for 2+ targets.

I have switched it to catching to be more efficient for 2+ targets. (There are more optimizations that can be done to optimize for many targets. I do not think it is worth it. I do not expect many targets to be the common case.)